ipa-core/src/protocol/context/dzkp_field.rs

```diff
@@ -119,87 +257,33 @@ impl DZKPBaseField for Fp61BitPrime {
-            .zip(x_right.data.iter())
-            .zip(prss_right.data.iter())
-            .zip(ac.data.iter())
-            .zip(e.data.iter())
-            .zip(bd.data.iter())
-            .map(|(((((((a, b), c), d), f), ac), e), bd)| {
-                (
-                    // g polynomial
-                    a.view_bits::<Lsb0>()
-                        .iter()
-                        .zip(c.view_bits::<Lsb0>().iter())
-                        .zip(e.view_bits::<Lsb0>().iter())
-                        .zip(ac.view_bits::<Lsb0>().iter())
-                        .flat_map(|(((a, c), e), ac)| {
-                            let one_minus_two_e = Fp61BitPrime::ONE
-                                + Fp61BitPrime::MINUS_TWO * Fp61BitPrime::from_bit(*e);
-                            [
-                                // g1=-2ac(1-2e),
-                                Fp61BitPrime::MINUS_TWO
-                                    * Fp61BitPrime::from_bit(*ac)
-                                    * one_minus_two_e,
-                                // g2=c(1-2e),
-                                Fp61BitPrime::from_bit(*c) * one_minus_two_e,
-                                // g3=a(1-2e),
-                                Fp61BitPrime::from_bit(*a) * one_minus_two_e,
-                                // g4=-1/2(1-2e),
-                                Fp61BitPrime::MINUS_ONE_HALF * one_minus_two_e,
-                            ]
-                        })
-                        .collect::<[Fp61BitPrime; BLOCK_SIZE]>(),
-                    // h polynomial
-                    b.view_bits::<Lsb0>()
-                        .iter()
-                        .zip(d.view_bits::<Lsb0>().iter())
-                        .zip(f.view_bits::<Lsb0>().iter())
-                        .zip(bd.view_bits::<Lsb0>().iter())
-                        .flat_map(|(((b, d), f), bd)| {
-                            let one_minus_two_f = Fp61BitPrime::ONE
-                                + Fp61BitPrime::MINUS_TWO * Fp61BitPrime::from_bit(*f);
-                            [
-                                // h1=bd(1-2f),
-                                Fp61BitPrime::from_bit(*bd) * one_minus_two_f,
-                                // h2=d(1-2f),
-                                Fp61BitPrime::from_bit(*d) * one_minus_two_f,
-                                // h3=b(1-2f),
-                                Fp61BitPrime::from_bit(*b) * one_minus_two_f,
-                                // h4=1-2f,
-                                one_minus_two_f,
-                            ]
-                        })
-                        .collect::<[Fp61BitPrime; BLOCK_SIZE]>(),
-                )
-            })
-            .collect::<Vec<_>>()
+        let f = prss_right;
+
+        let g = convert_values(a, c, &e, &TABLE_RIGHT);
+        let h = convert_values(b, d, f, &TABLE_LEFT);
+
+        zip(g.chunks_exact(BLOCK_SIZE), h.chunks_exact(BLOCK_SIZE))
```

Comment:
Do you mean pre-allocating the output vector instead of doing `collect` within `convert_prover`? Or are you suggesting some change to the way `convert_prover` is called? Or to how `convert_values` is called?

Reply:
I was thinking about this code https://github.com/private-attribution/ipa/blob/48407bbba2b0bfada99b28377ffa22d64012894f/ipa-core/src/protocol/context/dzkp_validator.rs#L477 but I don't think my idea works there. It could still be possible to implement `convert_prover_iter(..) -> impl Iterator<Item = UVTupleBlock>` as `convert_prover` is borrowing